File: backup/cyclesignal_flutter/lib/logic/gp_data.dart
import '../models/cycle_entry.dart';
import '../models/pattern.dart';
import 'cycle.dart';

class SymptomRow {
  final String name;
  final int cycles;
  final String window;
  final String avgSeverity;
  SymptomRow(this.name, this.cycles, this.window, this.avgSeverity);
}

class GPData {
  final int cyclesTracked;
  final int avgLength;
  final String range;
  final int cyclesOver35;
  final int missedPeriods;
  final List<SymptomRow> symptomRows;
  final String functionalText;
  final List<String> pcosPoints;
  final List<String> pmddPoints;
  final bool redFlag;
  GPData({
    required this.cyclesTracked,
    required this.avgLength,
    required this.range,
    required this.cyclesOver35,
    required this.missedPeriods,
    required this.symptomRows,
    required this.functionalText,
    required this.pcosPoints,
    required this.pmddPoints,
    required this.redFlag,
  });
}

int _cyclesWith(List<Cycle> cs, bool Function(CycleEntry) pred) =>
    cs.where((c) => c.entries.any(pred)).length;

String _window(List<Cycle> cs, bool Function(CycleEntry) pred) {
  final days = <int>[];
  for (final c in cs) {
    for (final e in c.entries) {
      if (pred(e)) days.add(e.cycleDay);
    }
  }
  if (days.isEmpty) return '—';
  final lo = days.reduce((a, b) => a < b ? a : b);
  final hi = days.reduce((a, b) => a > b ? a : b);
  return lo == hi ? 'Day $lo' : 'Days $lo–$hi';
}

String _avg(List<Cycle> cs, int? Function(CycleEntry) get) {
  final vals = <int>[];
  for (final c in cs) {
    for (final e in c.entries) {
      final v = get(e);
      if (v != null) vals.add(v);
    }
  }
  if (vals.isEmpty) return '—';
  return (vals.reduce((a, b) => a + b) / vals.length).toStringAsFixed(1);
}

GPData computeGPData(List<CycleEntry> entries, List<Cycle> cycles, PatternFlags flags) {
  final recent = cycles.length > 6 ? cycles.sublist(cycles.length - 6) : cycles;
  final total = recent.length;

  final rows = <SymptomRow>[];
  void push(String name, bool Function(CycleEntry) pred, [int? Function(CycleEntry)? sev]) {
    final n = _cyclesWith(recent, pred);
    if (n > 0) {
      rows.add(SymptomRow(name, n, _window(recent, pred), sev != null ? _avg(recent, sev) : '—'));
    }
  }

  push('Low / unstable mood', (e) => e.mood?.present == true, (e) => e.mood?.severity);
  push('Pain', (e) => e.physical.pain?.present == true, (e) => e.physical.pain?.severity);
  push('Fatigue', (e) => e.physical.fatigue?.present == true, (e) => e.physical.fatigue?.severity);
  push('Disrupted sleep', (e) => e.physical.sleep?.disrupted == true);
  push('Acne', (e) => e.physical.acne);
  push('Excess hair growth', (e) => e.physical.hairGrowth);
  push('Bloating', (e) => e.physical.bloating);
  push('Bleeding', (e) => e.bleeding.started);

  final fiCycles = _cyclesWith(recent, (e) => e.functionalImpact.present);
  final functionalText =
      'The user reported missing work or cancelling plans in $fiCycles of $total cycles due to symptoms.';
  final missed = recent.where((c) => !c.entries.any((e) => e.bleeding.started)).length;

  final pcos = <String>[];
  if (flags.pcosSignal) {
    if (flags.irregularCycles) {
      pcos.add('Irregular cycles (range ${flags.cycleLengthRange}, ${flags.cyclesOver35Days} over 35 days)');
    }
    if (flags.acneCycles > 0) pcos.add('Acne logged in ${flags.acneCycles} cycles');
    if (_cyclesWith(recent, (e) => e.physical.hairGrowth) > 0) {
      pcos.add('Excess hair growth reported');
    }
    if (flags.fatigueCycles > 0) pcos.add('Fatigue logged in ${flags.fatigueCycles} cycles');
  }

  final pmdd = <String>[];
  if (flags.pmddSignal) {
    pmdd.add(
        'Premenstrual mood symptoms in ${flags.premenstrualMoodCycles} of $total cycles (${flags.premenstrualWindowLabel})');
    pmdd.add('Mood symptoms reported to ease after bleeding begins');
    pmdd.add('Functional impact (missed work / cancelled plans) in $fiCycles of $total cycles');
  }

  return GPData(
    cyclesTracked: total,
    avgLength: flags.averageCycleLength,
    range: flags.cycleLengthRange,
    cyclesOver35: flags.cyclesOver35Days,
    missedPeriods: missed,
    symptomRows: rows,
    functionalText: functionalText,
    pcosPoints: pcos,
    pmddPoints: pmdd,
    redFlag: entries.any((e) => e.safetyFlagged),
  );
}

File: backup/cyclesignal_rn/app/src/lib/gpData.ts
import { CycleEntry, PatternFlags } from '../types';
import { Cycle } from './cycle';

export type SymptomRow = {
  name: string;
  cycles: number;
  window: string;
  avgSeverity: string;
};

export type GPData = {
  overview: {
    cyclesTracked: number;
    avgLength: number;
    range: string;
    cyclesOver35: number;
    missedPeriods: number;
  };
  symptomRows: SymptomRow[];
  functionalText: string;
  pcosPoints: string[];
  pmddPoints: string[];
  redFlag: boolean;
};

function cyclesWith(cycles: Cycle[], pred: (e: CycleEntry) => boolean): Cycle[] {
  return cycles.filter((c) => c.entries.some(pred));
}

function dayWindow(cycles: Cycle[], pred: (e: CycleEntry) => boolean): string {
  const days: number[] = [];
  for (const c of cycles) for (const e of c.entries) if (pred(e)) days.push(e.cycleDay);
  if (!days.length) return '—';
  const lo = Math.min(...days);
  const hi = Math.max(...days);
  return lo === hi ? `Day ${lo}` : `Days ${lo}–${hi}`;
}

function avgSeverity(cycles: Cycle[], get: (e: CycleEntry) => number | null): string {
  const vals: number[] = [];
  for (const c of cycles)
    for (const e of c.entries) {
      const v = get(e);
      if (v != null) vals.push(v);
    }
  if (!vals.length) return '—';
  return (vals.reduce((a, b) => a + b, 0) / vals.length).toFixed(1);
}

export function computeGPData(
  entries: CycleEntry[],
  cycles: Cycle[],
  flags: PatternFlags
): GPData {
  const recent = cycles.slice(-6);
  const total = recent.length;

  const rows: SymptomRow[] = [];
  const push = (name: string, pred: (e: CycleEntry) => boolean, sev?: (e: CycleEntry) => number | null) => {
    const n = cyclesWith(recent, pred).length;
    if (n > 0)
      rows.push({
        name,
        cycles: n,
        window: dayWindow(recent, pred),
        avgSeverity: sev ? avgSeverity(recent, sev) : '—',
      });
  };

  push('Low / unstable mood', (e) => !!e.mood?.present, (e) => e.mood?.severity ?? null);
  push('Pain', (e) => !!e.physical.pain?.present, (e) => e.physical.pain?.severity ?? null);
  push('Fatigue', (e) => !!e.physical.fatigue?.present, (e) => e.physical.fatigue?.severity ?? null);
  push('Disrupted sleep', (e) => !!e.physical.sleep?.disrupted);
  push('Acne', (e) => e.physical.acne);
  push('Excess hair growth', (e) => e.physical.hairGrowth);
  push('Bloating', (e) => e.physical.bloating);
  push('Bleeding', (e) => e.bleeding.started);

  const fiCycles = cyclesWith(recent, (e) => e.functionalImpact.present).length;
  const functionalText = `The user reported missing work or cancelling plans in ${fiCycles} of ${total} cycles due to symptoms.`;

  const missedPeriods = recent.filter((c) => !c.entries.some((e) => e.bleeding.started)).length;

  const pcosPoints: string[] = [];
  if (flags.pcosSignal) {
    if (flags.irregularCycles)
      pcosPoints.push(`Irregular cycles (range ${flags.cycleLengthRange}, ${flags.cyclesOver35Days} over 35 days)`);
    if (flags.acneCycles > 0) pcosPoints.push(`Acne logged in ${flags.acneCycles} cycles`);
    if (cyclesWith(recent, (e) => e.physical.hairGrowth).length > 0)
      pcosPoints.push('Excess hair growth reported');
    if (flags.fatigueCycles > 0) pcosPoints.push(`Fatigue logged in ${flags.fatigueCycles} cycles`);
  }

  const pmddPoints: string[] = [];
  if (flags.pmddSignal) {
    pmddPoints.push(
      `Premenstrual mood symptoms in ${flags.premenstrualMoodCycles} of ${total} cycles (${flags.premenstrualWindowLabel})`
    );
    pmddPoints.push('Mood symptoms reported to ease after bleeding begins');
    pmddPoints.push(`Functional impact (missed work / cancelled plans) in ${fiCycles} of ${total} cycles`);
  }

  const redFlag = entries.some((e) => e.safetyFlagged);

  return {
    overview: {
      cyclesTracked: total,
      avgLength: flags.averageCycleLength,
      range: flags.cycleLengthRange,
      cyclesOver35: flags.cyclesOver35Days,
      missedPeriods,
    },
    symptomRows: rows,
    functionalText,
    pcosPoints,
    pmddPoints,
    redFlag,
  };
}
